### PR TITLE
Use explicit torch.autograd.grad insted of .backward

### PR DESCRIPTION
**Pull request by Balandat**
Being explicit in computing the gradients that we want seems to be the right thing to do.

### CODE REVIEW

**Comment by eytan** [comment]
Any context on this change? It does make the code more complicated / less idiomatic.

**Comment by Balandat** [comment]
Mostly b/c it gets rid of some gotchas. 

Using `backward` means you always have to be aware where you're calling it, what you're calling it on, whether you've properly required gradients for the variables that you're differentiating w.r.t. You also need to remember to zero out your gradients if you reuse the tensors. Using `grad` is less gotcha-esque, and also more functional which makes it easier to understand what's going on b/c all the relevant code is in one place (rather than distributed into a `backward` call, an accessing of the `.grad` property of the relevant tensors, and zeroing out gradients). 

gotcha 1:
```
X = torch.rand(3)
f(X).backward()
X.grad # None
```
```
X = torch.rand(3)
torch.autograd.grad(f(X), X)  # raises RuntimeError: One of the differentiated Tensors does not require grad
```

gotcha2:
```
X = torch.rand(3, requires_grad=True)

Y = X ** 2
Y.sum().backward()
X.grad  # tensor([1.4779, 0.6099, 1.1670])

Y = X ** 2
Y.sum().backward()
X.grad # tensor([2.9558, 1.2199, 2.3339])
```

```
X = torch.rand(3, requires_grad=True)

Y = X ** 2
torch.autograd.grad(Y.sum(), X)[0]  # tensor([1.4779, 0.6099, 1.1670])

Y = X ** 2
torch.autograd.grad(Y.sum(), X)[0]  # tensor([1.4779, 0.6099, 1.1670])
```

File: botorch/acquisition/objective.py [hunk 1]
```diff
@@ -63,7 +63,7 @@ class LinearMCObjective(MCAcquisitionObjective):
     `mc_obj(samples) = sum_{i} weights[i] * samples[..., i]`
 
     Example:
-        >>> # example for a two outcomes
+        Example for a model with two outcomes:
```
Comment (sdaulton) [inline botorch/acquisition/objective.py]: Sphinx won't correctly parse the following lines as a code block. Here are few options that will be correctly parsed: add a blank line after L66 before the code block, make this line a comment (starting with `>>> #`), or indent the code block to the next level.

File: botorch/gen.py [hunk 2]
```diff
@@ -101,9 +101,8 @@ def f(x):
         )
         X_fix = fix_features(X=X, fixed_features=fixed_features)
         loss = -acquisition_function(X_fix).sum()
-        loss.backward()
+        gradf = _arrayify(torch.autograd.grad(loss, X)[0].contiguous().view(-1))
```
Comment (sdaulton) [inline botorch/gen.py]: Mind adding a comment for posterity, since `backward` is more common.

**Pull request closed without merging**